sway/input/keyboard.c

```diff
@@ -534,20 +534,58 @@ static void handle_xkb_context_log(struct xkb_context *context,
 
 struct xkb_keymap *sway_keyboard_compile_keymap(struct input_config *ic,
 		char **error) {
-	struct xkb_rule_names rules = {0};
-	if (ic) {
-		input_config_fill_rule_names(ic, &rules);
-	}
-
 	struct xkb_context *context = xkb_context_new(XKB_CONTEXT_NO_FLAGS);
 	if (!sway_assert(context, "cannot create XKB context")) {
 		return NULL;
 	}
 	xkb_context_set_user_data(context, error);
 	xkb_context_set_log_fn(context, handle_xkb_context_log);
 
-	struct xkb_keymap *keymap =
-		xkb_keymap_new_from_names(context, &rules, XKB_KEYMAP_COMPILE_NO_FLAGS);
+	struct xkb_keymap *keymap = NULL;
+
+	if (ic && ic->xkb_file) {
+		FILE *keymap_file = fopen(ic->xkb_file, "r");
+		if (!keymap_file) {
+			if (error) {
+				size_t len = snprintf(NULL, 0, "cannot read XKB file %s: %s",
+						ic->xkb_file, strerror(errno)) + 1;
+				*error = malloc(len);
+				if (*error) {
+					snprintf(*error, len, "cannot read XKB file %s: %s",
+							ic->xkb_file, strerror(errno));
+				} else {
+					sway_log_errno(SWAY_ERROR, "cannot read XKB file %s: %s",
+							ic->xkb_file, strerror(errno));
+				}
+			}
+		}
+
+		keymap = xkb_keymap_new_from_file(context, keymap_file,
```

Comment:
If `keymap_file` is `NULL`, this shouldn't continue, but does need to cleanup. The most readable option would be to add a `cleanup:` label before the `xkb_context_set_user_data` line and add a `goto cleanup` to the end of the `if` above.

Reply:
Do you think that if keymap_file is null that we should default to the previous setup with rules?

Reply:
This would be cleaner
`if (!keymap_file) {
 //handle error
} else {
  keymap = xkb_keymap_new_from_file(...)
  //close file
}
if (!keymap) {
 // rule setup
}
//context and return`

Reply:
No. One of the cases where this gets called is when storing the delta input config. If `NULL` is returned from this function, the delta input config is rejected for producing an invalid keymap. If this gets changed to return a valid keymap (from the rules), it'll result in the delta config being applied anyway and `*error` being leaked.